pkg/controller/controller.go

```diff
@@ -125,9 +241,286 @@ func (c *Controller) Start(wg *sync.WaitGroup) error {
+
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			c.workQueue.Forget(obj)
+			runtime.HandleError(fmt.Errorf("expected string from queue, got %#v", obj))
+			return nil
+		}
+
+		if err := c.handleClusterUpdate(key); err != nil {
+			return fmt.Errorf("error syncing cluster '%s': %v", key, err)
+		}
+
+		c.workQueue.Forget(obj)
+		c.logger.Info("successfully synced item", zap.String("key", key))
+
+		return nil
+	}(obj)
+
+	if err != nil {
+		runtime.HandleError(err)
+	}
+
+	return true
+}
+
+func (c *Controller) handleClusterUpdate(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	cluster, err := c.clusterLister.M3DBClusters(namespace).Get(name)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			// give up processing if this cluster doesn't exist
+			runtime.HandleError(fmt.Errorf("clusters '%s' no longer exists", key))
+			return nil
+		}
+
+		return err
+	}
+
+	// MUST create a deep copy of the cluster or risk corruping cache! Technically
+	// only need if we modify, but we frequently do that so let's deep copy to
+	// start and remove unnecessary calls later to optimize if we want.
+	cluster = cluster.DeepCopy()
+
+	// this deepcopy is probably unnecessary
+	isoGroups := cluster.DeepCopy().Spec.IsolationGroups
+
+	if len(isoGroups) == 0 {
+		// nothing to do, no groups to create in
+		return nil
+	}
+
+	zones := make([]string, len(isoGroups))
+	for i := range zones {
+		zones[i] = isoGroups[i].Name
+	}
+
+	childrenSets, err := c.getChildStatefulSets(cluster)
+	if err != nil {
+		return err
+	}
+
+	for _, sts := range childrenSets {
+		// if any of the statefulsets aren't ready, wait until they are as we'll get
+		// another event (ready == bootstrapped)
+		if sts.Spec.Replicas != nil && *sts.Spec.Replicas != sts.Status.ReadyReplicas {
+			// TODO(schallert): figure out what to do if replicas is not set
+			c.logger.Info("waiting for statefulset to be ready", zap.String("name", sts.Name), zap.Int32("ready", sts.Status.ReadyReplicas))
+			return nil
+		}
+	}
+
+	// At this point all existing statefulsets are bootstrapped.
+
+	if len(childrenSets) != len(zones) {
+		nextID := len(childrenSets)
+		// create a statefulset
+		selfRef := metav1.NewControllerRef(cluster, schema.GroupVersionKind{
+			Group:   myspec.SchemeGroupVersion.Group,
+			Version: myspec.SchemeGroupVersion.Version,
+			// TODO(schallert): use a const here
+			Kind: "m3dbcluster",
+		})
+
+		name := fmt.Sprintf("%s-%d", cluster.Name, nextID)
+		sts := newBaseStatefulSet(name, cluster.Spec.Image, cluster.Name, zones[nextID], 1)
+
+		// Create an Affinity policy for isolationgroup.
+		nodeAffinity := &corev1.NodeAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
+				NodeSelectorTerms: []corev1.NodeSelectorTerm{
+					{
+						MatchExpressions: []corev1.NodeSelectorRequirement{
```

Comment:
Can this be made local minikube friendly ? Or update the `README` to explain that a label is needed for the minikube?

```
kubectl label nodes minikube failure-domain.beta.kubernetes.io/zone=<zone_name>
```

Reply:
Yeah I'll update the README